crates/primitives/transactions/src/compute_hash.rs

```diff
@@ -1,37 +1,85 @@
+use starknet_api::transaction::{
+    Calldata, DeclareTransaction, DeclareTransactionV0V1, DeclareTransactionV2, DeclareTransactionV3,
+    DeployAccountTransaction, DeployAccountTransactionV1, DeployAccountTransactionV3, DeployTransaction,
+    InvokeTransaction, InvokeTransactionV0, InvokeTransactionV1, InvokeTransactionV3, L1HandlerTransaction, Resource,
+    ResourceBoundsMapping, Transaction, TransactionHash,
+};
 use starknet_core::crypto::compute_hash_on_elements;
 use starknet_core::utils::starknet_keccak;
 use starknet_crypto::FieldElement;
 
-use super::{
-    DeclareTransaction, DeclareTransactionV0, DeclareTransactionV1, DeclareTransactionV2, DeployAccountTransaction,
-    HandleL1MessageTransaction, InvokeTransaction, InvokeTransactionV0, InvokeTransactionV1, Transaction,
-    UserTransaction, SIMULATE_TX_VERSION_OFFSET,
-};
-use crate::{DeployTransaction, UserOrL1HandlerTransaction, LEGACY_BLOCK_NUMBER, LEGACY_L1_HANDLER_BLOCK};
+use super::SIMULATE_TX_VERSION_OFFSET;
+use crate::LEGACY_BLOCK_NUMBER;
 
 const DECLARE_PREFIX: &[u8] = b"declare";
 const DEPLOY_ACCOUNT_PREFIX: &[u8] = b"deploy_account";
 const DEPLOY_PREFIX: &[u8] = b"deploy";
 const INVOKE_PREFIX: &[u8] = b"invoke";
 const L1_HANDLER_PREFIX: &[u8] = b"l1_handler";
+const L1_GAS: &[u8] = b"L1_GAS";
+const L2_GAS: &[u8] = b"L2_GAS";
 
 pub trait ComputeTransactionHash {
     fn compute_hash<H: HasherT>(
         &self,
         chain_id: Felt252Wrapper,
         offset_version: bool,
         block_number: Option<u64>,
-    ) -> Felt252Wrapper;
+    ) -> TransactionHash;
+}
+
+fn convert_calldata(calldata: Calldata) -> Vec<FieldElement> {
+    calldata.0.iter().map(|f| Felt252Wrapper::from(*f).into()).collect()
 }
 
-fn convert_calldata(data: &[Felt252Wrapper]) -> &[FieldElement] {
-    // Non-copy but less dangerous than transmute
-    // https://doc.rust-lang.org/std/mem/fn.transmute.html#alternatives
-    unsafe { core::slice::from_raw_parts(data.as_ptr() as *const FieldElement, data.len()) }
+fn prepare_resource_bound_value(resource_bounds_mapping: &ResourceBoundsMapping, resource: Resource) -> FieldElement {
+    let mut buffer = [0u8; 32];
+    buffer[2..8].copy_from_slice(match resource {
+        Resource::L1Gas => L1_GAS,
+        Resource::L2Gas => L2_GAS,
+    });
+    if let Some(resource_bounds) = resource_bounds_mapping.0.get(&resource) {
+        buffer[8..16].copy_from_slice(&resource_bounds.max_amount.to_be_bytes());
+        buffer[16..].copy_from_slice(&resource_bounds.max_price_per_unit.to_be_bytes());
+    };
+
+    // Safe to unwrap because we left most significant bit of the buffer empty
+    FieldElement::from_bytes_be(&buffer).unwrap()
+}
+
+fn prepare_data_availability_modes(
+    nonce_data_availability_mode: DataAvailabilityMode,
+    fee_data_availability_mode: DataAvailabilityMode,
+) -> FieldElement {
+    let mut buffer = [0u8; 32];
+    buffer[64..96].copy_from_slice(&(nonce_data_availability_mode as u32).to_be_bytes());
```

Comment:
Same thing here, what data is this actually collecting and why?

Reply:
change